### DISCUSSION TITLE
How to reduce time to generate .UF2 with frozen script

### DISCUSSION THREAD

**Discussion by rcalcagno** [micropython/micropython#13405] [Core Development] [answered]
Good evening. As part of my project I need to freeze a relative large (+3000 lines) .py script for a Raspberry Pi Pico. As the script grew too big, and I still need to add more features, I find myself in trouble since I can't use the interpreter anymore; it keeps throwing many errors. I found that freezing my routine into the firmware allows me to keep going. 

So, I setup a Docker container that builds all the neccesary tools, and I can generate the .UF2 firmware with the required modifications, including the single module "mymodule.py" on  /ports/rp2/modules that gets compiled and can be imported and run, so far so good.

Thing is, as expected, every single modification to "mymodule.py" involves generating ALL the modules that make up MicroPython, even though none of them have changed, except for my single file. This takes a very long amount of time to process; I was able to reduce it somehow by using a multi-stage container; in which the first stage generates up to ```mpy-cross```; and the second stage takes from there and generates the .UF2. But even this is rather time consuming.

Is there any way of configuring CMake to stop right when it finished generating all the .obj files pertinent to MicroPython, ideally on the first stage; then compile my single file, and finally linking all together to generate the .UF2?

Any help will be much appreciated. Sincerely,

Ricardo.

**Comment by robert-hh** [comment, marked as answer]
That could be related to the docker build style. I use a standard file system build environment, and If I modify a frozen script after build and rebuild the firmware, then just that modified file is re-compiled and the binary is created.

**Reply by rcalcagno** [reply]
Dear Robert: can't thank you enough. That was exactly what was happening; apparently CMake has joined the ever-growing list of apps and programs that are smarter than me :) (wait until I stumble upon ChatGPT!). 

I tried a "baremetal" installation of Ubuntu and it was exactly as you said; the script is clever enough to compile and link only the parts that changed. In the compilation line however, as the instructions say should be ```make submodules && make clean && make```,  had to get rid of the  ```make clean``` though, as it erases all the intermediate steps which are neccesary in this case. With this news, I can go back to my docker and try again; I believe it will work.

Once again, thanks a lot for your help. Sincerely,

Ricardo

**Comment by rcalcagno** [comment]
Building with the proper filesystem takes care of the unmodified components
